cuda_core/cuda/core/experimental/_launcher.py

```diff
@@ -86,3 +89,17 @@ def launch(stream, config, kernel, *kernel_args):
                 int(kernel._handle), *config.grid, *config.block, config.shmem_size, stream.handle, args_ptr, 0
             )
         )
+
+
+def _check_cooperative_launch(kernel: Kernel, config: LaunchConfig, stream: Stream):
+    dev = stream.device
+    num_sm = dev.properties.multiprocessor_count
+    max_grid_size = (
+        kernel.occupancy.max_active_blocks_per_multiprocessor(_reduce_tuple(config.block), config.shmem_size) * num_sm
+    )
+    if _reduce_tuple(config.grid) > max_grid_size:
+        # For now let's try not to be smart and adjust the grid size behind users' back.
+        # We explicitly ask users to adjust.
+        raise ValueError(
+            "The specified grid size ({} * {} * {}) exceeds the limit ({}).".format(*config.grid, max_grid_size)
```

Comment:
nitpick: use f-string instead of `str.format`?

Reply:
done in commit 138f0d1